src/core/exec-invoke.c

```diff
@@ -1178,64 +1246,8 @@ static int setup_pam(
-
-                assert_se(sigprocmask_many(SIG_BLOCK, NULL, SIGTERM) >= 0);
-
-                /* The child's job is to reset the PAM session on termination */
-                barrier_set_role(&barrier, BARRIER_CHILD);
-
-                /* Make sure we don't keep open the passed fds in this child. We assume that otherwise only
-                 * those fds are open here that have been opened by PAM. */
-                (void) close_many(fds, n_fds);
-
-                /* Also close the 'exec_fd' in the child, since the service manager waits for the EOF induced
-                 * by the execve() to wait for completion, and if we'd keep the fd open here in the child
-                 * we'd never signal completion. */
-                exec_fd = safe_close(exec_fd);
-
-                /* Drop privileges - we don't need any to pam_close_session and this will make
-                 * PR_SET_PDEATHSIG work in most cases.  If this fails, ignore the error - but expect sd-pam
-                 * threads to fail to exit normally */
-
-                r = fully_set_uid_gid(uid, gid, /* supplementary_gids= */ NULL, /* n_supplementary_gids= */ 0);
-                if (r < 0)
-                        log_warning_errno(r, "Failed to drop privileges in sd-pam: %m");
-
-                (void) ignore_signals(SIGPIPE);
-
-                /* Tell the parent that our setup is done. This is especially important regarding dropping
-                 * privileges. Otherwise, unit setup might race against our setresuid(2) call.
-                 *
-                 * If the parent aborted, we'll detect this below, hence ignore return failure here. */
-                (void) barrier_place(&barrier);
-
-                /* Check if our parent process might already have died? */
-                if (getppid() == parent_pid.pid) {
-                        sigset_t ss;
-                        int sig;
-
-                        assert_se(sigemptyset(&ss) >= 0);
-                        assert_se(sigaddset(&ss, SIGTERM) >= 0);
-
-                        assert_se(sigwait(&ss, &sig) == 0);
-                        assert(sig == SIGTERM);
-                }
-
-                /* If our parent died we'll end the session */
-                if (getppid() != parent_pid.pid) {
-                        pam_code = pam_close_session_and_delete_credentials(handle, flags);
-                        if (pam_code != PAM_SUCCESS)
-                                goto child_finish;
-                }
-
-                ret = 0;
-
-        child_finish:
-                /* NB: pam_end() when called in child processes should set PAM_DATA_SILENT to let the module
-                 * know about this. See pam_end(3) */
-                (void) pam_end(handle, pam_code | flags | PAM_DATA_SILENT);
-                _exit(ret);
+                r = pam_child_main(handle, flags, &barrier, fds, n_fds, exec_fd, uid, gid, &parent_pid);
+                _exit(r != 0 ? EXIT_PAM : EXIT_SUCCESS);
```

Comment:
pam_child_main now returns a pam_code, so I think this is clearer if the comparison is explicit (like before):
```C
r = pam_child_main(…);
_exit(r == PAM_SUCCESS ? EXIT_SUCCESS : EXIT_PAM);
```